asyncio_run_in_process/run_in_process.py

```diff
@@ -113,22 +118,31 @@ async def _monitor_state(
             next_expected_state = State(proc.state + 1)
             logger.debug(
                 "Waiting for next expected state (%s) from child (%s)", next_expected_state, proc)
-            child_state_as_byte = await loop.run_in_executor(None, read_exactly, from_child, 1)
+            try:
+                child_state_as_byte = await loop.run_in_executor(None, read_exactly, from_child, 1)
+            except asyncio.CancelledError:
+                # When the sub process is sent a SIGKILL, the write end of the pipe used in
+                # read_exactly is never closed and the thread above attempting to read from it
+                # will prevent us from leaving this fdopen() context, so we need to close the
+                # write end ourselves to ensure the read_exactly() returns.
+                logger.debug(
+                    "_monitor_state() cancelled while waiting data from child, closing "
+                    "child_write_fd to ensure we exit")
+                os.close(child_write_fd)
+                raise
 
             try:
                 child_state = State(ord(child_state_as_byte))
             except TypeError:
                 raise InvalidState(f"Child sent state: {child_state_as_byte.hex()}")
 
             if not proc.state.is_next(child_state):
-                if proc.state is State.FINISHED:
-                    # This case covers when the process is killed with a SIGKILL
-                    # and jumps directly to the finished state.
-                    return
```

Comment:
This block was only reachable because `test_open_in_proc_SIGKILL_while_running` was calling kill() before the sub proc had reached the EXECUTING state, but on py37 the SIGKILL was received only after the sub proc had sent us the EXECUTING state (causing the run_in_executor(readexactly) to complete instead of being cancelled), then hit the if above and _monitor_state() to return (which in this case works because there's no pending read like when the run_in_executor(readexactly) is cancelled. On py38 that test was hanging forever because the stars were no longer all lined up in the exact sequence it needed